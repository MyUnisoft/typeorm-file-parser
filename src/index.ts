// Import Node.js Dependencies
import { createReadStream } from "fs";

// Import Third-party Dependencies
import split2 from "split2";

// Import Internal Dependencies
import { parseDecorator, TypeORMDecoratorBase, TypeORMDecoratorExtended } from "./decorator/parser";

// CONSTANTS
const kArrobaseChar = "@";
const kSlashChar = "/";

export async function* lazyFetchFileDecorator(fileLocation: string) {
  const rStream = createReadStream(fileLocation).pipe(split2());
  let pastConcat: string | null = null;

  for await (const line of rStream) {
    const trimedLine = line.trim() as string;
    if (trimedLine.includes("export interface")) {
      break;
    }

    const isCommentLine = trimedLine.startsWith(kSlashChar);
    if (isCommentLine) {
      continue;
    }

    const fieldLineResult = /^([a-zA-Z0-9_]+)\s*:\s*([a-zA-Z]+)/g.exec(trimedLine);
    if (fieldLineResult !== null) {
      const [, name, value] = fieldLineResult;
      yield [name, value];

      continue;
    }

    const isEndLine = trimedLine.charAt(trimedLine.length - 1) === ")";
    const isStartLine = trimedLine.startsWith(kArrobaseChar);

    if (isEndLine) {
      const finalPayload = isStartLine ? trimedLine : pastConcat + trimedLine;
      yield parseDecorator(finalPayload);

      pastConcat = null;
      continue;
    }

    pastConcat = (pastConcat || "") + trimedLine;
  }
}

type DecoratorExWithoutName = Omit<TypeORMDecoratorExtended, "name"> | string;

function removeDecoratorName(decorator: TypeORMDecoratorBase): DecoratorExWithoutName {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { name, ...otherProperties } = decorator;
  if (Object.keys(otherProperties).length === 0) {
    return name;
  }

  return otherProperties;
}

export interface TypeORMProperty {
  type: string;
  decorators: Record<string, TypeORMDecoratorBase>;
}

export interface ParsedTypeORMResult {
  unique?: DecoratorExWithoutName;
  properties: Record<string, TypeORMProperty>;
}

export async function readFile(fileLocation: string): Promise<ParsedTypeORMResult> {
  let memoryColumns: TypeORMDecoratorBase[] = [];

  const result: ParsedTypeORMResult = {
    properties: {}
  };
  const asyncIter = await lazyFetchFileDecorator(fileLocation);
  await asyncIter.next();

  for await (const decorator of asyncIter) {
    if (Array.isArray(decorator)) {
      const decorators = {};
      for (const decorator of memoryColumns) {
        decorators[decorator.name] = decorator;
      }

      const [propertyName, propertyValue] = decorator;
      result.properties[propertyName] = {
        type: propertyValue,
        decorators
      };

      memoryColumns = [];
    }
    else if (decorator.name === "Unique") {
      result.unique = removeDecoratorName(decorator);
    }
    else {
      memoryColumns.push(decorator);
    }
  }

  return result;
}
