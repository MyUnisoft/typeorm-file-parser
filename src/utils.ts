export function charSet(...plages: (string | [number, number])[]): Set<string> {
  const ret = new Set<string>();

  for (const plage of plages) {
    if (typeof plage === "string") {
      ret.add(plage);
      continue;
    }

    for (let tid = plage[0]; tid <= plage[1]; tid++) {
      ret.add(String.fromCharCode(tid));
    }
  }

  return ret;
}

export const END_OF_SEQUENCE = Symbol("EOS");
export type SEQUENCE_OR_TOKEN<T> = typeof END_OF_SEQUENCE | T;

export function getNextItem<T>(iterator: IterableIterator<T>): SEQUENCE_OR_TOKEN<T> {
  const item = iterator.next();

  return item.done ? END_OF_SEQUENCE : item.value;
}
