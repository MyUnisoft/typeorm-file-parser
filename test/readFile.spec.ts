// Import Node.js Dependencies
import path from "path";

// Import Third-party Dependencies
import is from "@slimio/is";

// Import Internal Dependencies
import { readFile } from "../src/index";

// CONSTANTS
const kFixturesPath = path.join(__dirname, "fixtures");

describe("readFile", () => {
  it("should be able to read GiArticle", async() => {
    const result = await readFile(path.join(kFixturesPath, "GiArticle.ts"));

    expect(is.plainObject(result)).toStrictEqual(true);
    expect("properties" in result).toStrictEqual(true);
    expect("unique" in result).toStrictEqual(true);

    expect(result.unique).toMatchObject({
      constraintName: "GiArticle_reference",
      columns: ["reference"]
    });

    expect(result).toMatchSnapshot();
  });

  it("should be able to read Empty", async() => {
    const result = await readFile(path.join(kFixturesPath, "Empty.ts"));

    expect("unique" in result).toStrictEqual(false);

    const id = result.properties.id;
    expect(id.type).toStrictEqual("number");
    expect("PrimaryColumn" in id.decorators).toStrictEqual(true);
  });
});

