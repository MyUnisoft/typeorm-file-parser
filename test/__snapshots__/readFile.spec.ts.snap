// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readFile should be able to read GiArticle 1`] = `
Object {
  "properties": Object {
    "accountArticle": Object {
      "decorators": Object {
        "JoinColumn": Object {
          "name": "JoinColumn",
          "properties": Object {
            "name": "account_article_id",
          },
        },
        "OneToOne": Object {
          "name": "OneToOne",
          "properties": Object {},
          "table": "GiArticle",
          "tableColumn": "id",
        },
      },
      "type": "GiArticle",
    },
    "amount": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "enum": "giAmount",
          },
          "type": "enum",
        },
      },
      "type": "giAmount",
    },
    "archived": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "default": "false",
          },
          "type": "boolean",
        },
      },
      "type": "boolean",
    },
    "billed": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "default": "false",
          },
          "type": "boolean",
        },
      },
      "type": "boolean",
    },
    "clientAccount": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "length": 6,
          },
          "type": "varchar",
        },
      },
      "type": "string",
    },
    "clientDeductibleVat": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "length": 6,
          },
          "type": "varchar",
        },
      },
      "type": "string",
    },
    "clientVat": Object {
      "decorators": Object {
        "JoinColumn": Object {
          "name": "JoinColumn",
          "properties": Object {
            "name": "client_vat_id",
          },
        },
        "ManyToOne": Object {
          "name": "ManyToOne",
          "properties": Object {
            "nullable": false,
          },
          "table": "Vat",
          "tableColumn": "giArticlesClient",
        },
      },
      "type": "Vat",
    },
    "exercice": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "enum": "giExercice",
          },
          "type": "enum",
        },
      },
      "type": "string",
    },
    "firmAccount": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "length": 6,
          },
          "type": "varchar",
        },
      },
      "type": "number",
    },
    "firmCollectedVat": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "length": 6,
          },
          "type": "varchar",
        },
      },
      "type": "string",
    },
    "firmVat": Object {
      "decorators": Object {
        "JoinColumn": Object {
          "name": "JoinColumn",
          "properties": Object {
            "name": "firm_vat_id",
          },
        },
        "ManyToOne": Object {
          "name": "ManyToOne",
          "properties": Object {
            "nullable": false,
          },
          "table": "Vat",
          "tableColumn": "giArticlesFirm",
        },
      },
      "type": "Vat",
    },
    "firstDeadline": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {},
          "type": "smallint",
        },
      },
      "type": "number",
    },
    "id": Object {
      "decorators": Object {
        "PrimaryGeneratedColumn": Object {
          "name": "PrimaryGeneratedColumn",
        },
      },
      "type": "number",
    },
    "mandatoryArticle": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {},
          "type": "boolean",
        },
      },
      "type": "boolean",
    },
    "missions": Object {
      "decorators": Object {
        "JoinTable": Object {
          "name": "JoinTable",
          "properties": Object {
            "name": "gi_articles_missions",
          },
        },
        "ManyToMany": Object {
          "name": "ManyToMany",
          "properties": Object {},
          "table": "GiMission",
          "tableColumn": "articles",
        },
      },
      "type": "GiMission",
    },
    "name": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "length": 50,
          },
          "type": "varchar",
        },
      },
      "type": "string",
    },
    "note": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "nullable": true,
          },
          "type": "text",
        },
      },
      "type": "string",
    },
    "periodicity": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "enum": "periodicity",
          },
          "type": "enum",
        },
      },
      "type": "periodicity",
    },
    "reference": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "length": 6,
          },
          "type": "varchar",
        },
      },
      "type": "string",
    },
    "type": Object {
      "decorators": Object {
        "Column": Object {
          "name": "Column",
          "properties": Object {
            "enum": "giTypeArticle",
          },
          "type": "enum",
        },
      },
      "type": "giTypeArticle",
    },
    "vintages": Object {
      "decorators": Object {
        "OneToMany": Object {
          "name": "OneToMany",
          "properties": Object {
            "cascade": true,
          },
          "table": "GiVintage",
          "tableColumn": "article",
        },
      },
      "type": "GiVintage",
    },
    "workUnit": Object {
      "decorators": Object {
        "ManyToOne": Object {
          "name": "ManyToOne",
          "properties": Object {
            "cascade": true,
            "nullable": false,
          },
          "table": "WorkUnit",
          "tableColumn": "articles",
        },
      },
      "type": "WorkUnit",
    },
  },
  "unique": Object {
    "columns": Array [
      "reference",
    ],
    "constraintName": "GiArticle_reference",
  },
}
`;
